### DISCUSSION TITLE
equidistant_unchecked on a vector

### DISCUSSION THREAD

**Discussion by boehs** [NicolasKlenert/enterpolation#22] [Q&A] [answered]
I have a vector instead of an array because the length of my gradient (via palette) cannot be known at build time:

```rs
        let colors = input
            .split(",")
            .flat_map(|x| Srgb::from_str(x.trim_start()).map(|c| c.into_linear()))
            .collect::<Vec<_>>();
```

I'd like to generate a gradient from it, like is shown in the examples:

``` rs
let gradient = ConstEquidistantLinear::<f32, _, 3>::equidistant_unchecked([
    LinSrgb::new(0.00, 0.05, 0.20),
    LinSrgb::new(0.70, 0.10, 0.20),
    LinSrgb::new(0.95, 0.90, 0.30),
]);
```

but I need to use a vec. I tried:

```rs
Linear::builder().elements(self.0.clone()).equidistant::<f64>().normalized().build();
```

but the build function has the following error:

```rs
the method `build` exists for struct `LinearBuilder<Equidistant, Vec<Rgb<Linear<Srgb>>>, Identity, WithoutWeight>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`Rgb<stripper::primitives::color::encoding::Linear<Srgb>>: Merge
rgb.rs(71, 1): doesn't satisfy `_: Merge`
```

I also tried converting the vec to an array:

```rs
fn vec_to_arr<T, const N: usize>(v: Vec<T>) -> [T; N] {
    v.try_into()
        .unwrap_or_else(|v: Vec<T>| panic!("Expected a Vec of length {} but it was {}", N, v.len()))
}

let thing =
            ConstEquidistantLinear::<f32, _, 3>::equidistant_unchecked(vec_to_arr(self.0.clone()));
```

, but the length (3) is still hardcoded.

I see

> Linear equidistant constant interpolations are often wanted to define some specific curve (like a specific gradient). To create such interpolation, the builder pattern can not be used yet. Instead one should create a linear interpolation directly with the [equidistant_unchecked()](https://docs.rs/enterpolation/latest/enterpolation/linear/struct.Linear.html#method.equidistant_unchecked) constructor.

In the docs, but I don't know if it's relevant.

How should I do this correctly?

Thanks, Evan

**Comment by boehs** [comment]
sorry, I swear I've been at this for a hour before I opened this 😁 , but I finally dived into the code of this crate and came out with this, which works wonderfully. Do tell if there is a better way 🙂 .

```rs
        let thing = Linear::new_unchecked(
            self.0.clone(),
            Equidistant::<f32>::normalized(self.0.len()),
            Identity::new(),
        );
```

**Reply by NicolasKlenert** [reply, marked as answer]
Sorry, I did not see this disscussion until now! I tried to reproduce your error but I'm not able to. Code like below works for me. Did you import the `Merge` trait?

``` rust
use enterpolation::{Curve, Generator, Merge, linear::Linear};
use palette::{LinSrgb, IntoColor, Mix};

let elements = vec![
    LinSrgb::new(0.00, 0.05, 0.20),
    LinSrgb::new(0.70, 0.10, 0.20),
    LinSrgb::new(0.95, 0.90, 0.30),
];
let gradient = Linear::builder()
    .elements(elements)
    .equidistant::<f32>()
    .normalized()
    .build()
    .unwrap();
```

Otherwise `new_unchecked` is fine if you don't need to check if the curve is valid and your code looks good in that case!
